### PR TITLE
Optionally support aws-lc-rs

### PR DESCRIPTION
**Pull request by ctz**
This passes the same set of testing as *ring*. I haven't tested performance.

(I've removed some of the novel code-reuse via `super::` present in prior versions of this PR, because `aws-lc-rs` does not gate RSA on the `alloc` feature. That was a detail I think worth getting right, but means quite a bit of similarity between `aws_ls_rs_algs.rs` and `ring_algs.rs`. But I am using a similar trick to reuse the test code.)

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -95,3 +95,33 @@ pub use ring_algs::{
     RSA_PKCS1_3072_8192_SHA384, RSA_PSS_2048_8192_SHA256_LEGACY_KEY,
     RSA_PSS_2048_8192_SHA384_LEGACY_KEY, RSA_PSS_2048_8192_SHA512_LEGACY_KEY,
 };
+
+/// An array of all the verification algorithms exported by this crate.
+///
+/// This will be empty if the crate is built without the `ring` feature.
+pub static ALL_VERIFICATION_ALGS: &[&dyn SignatureVerificationAlgorithm] = &[
```
Comment (cpu) [inline src/lib.rs]: Thank you! It's been gross seeing variations of this replicating across various tests :sweat_smile:

**Review by cpu — COMMENTED** [review]
This looks great :rocket:

> Build currently known-broken on windows. This will be stuck as a draft until then.

Is there a path forward in the event Windows support in the upstream crate isn't prioritized?

> The first commit is somewhat odd/novel, to allow a single file to be used twice with different dependencies, resolved via super::lib. Interested in any alternatives to that!

It seems like a reasonable approach to me but I'm not sure I have the experience to see any lurking danger from the novelty. Was there a particular reason you were interested in alternatives that would be good to know about?

File: src/ring_algs.rs [hunk 2]
```diff
@@ -1,564 +1,15 @@
...
-// ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
-// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
-// ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
-// OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+use ring as lib;
 
-use crate::signed_data::{alg_id, InvalidSignature, SignatureVerificationAlgorithm};
-use ring::signature;
+#[path = "ringlike_algs.rs"]
+mod algs;
```
Comment (djc) [inline src/ring_algs.rs:4]: Why rename this module? Might as well call this `ringlike_algs`? Maybe call it `impls` to have more of a contrast from the `algs` module you also define?

File: src/ring_algs.rs [hunk 3]
```diff
@@ -1,564 +1,15 @@
...
-// copyright notice and this permission notice appear in all copies.
-//
-// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHORS DISCLAIM ALL WARRANTIES
-// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
-// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR
-// ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
-// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
-// ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
-// OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+use ring as lib;
```
Comment (djc) [inline src/ring_algs.rs:1]: I don't think we should alias this as just `lib`, that seems a little too generic.

File: src/ring_algs.rs [hunk 4]
```diff
@@ -1,564 +1,15 @@
-// Copyright 2015 Brian Smith.
-//
-// Permission to use, copy, modify, and/or distribute this software for any
```
Comment (djc) [inline src/ring_algs.rs:3]: Splitting the rename from `ring_algs.rs` to `ringlike_algs.rs` into a separate commit will make this a lot easier to review.

File: tests/signatures.rs [hunk 5]
```diff
@@ -56,17 +56,17 @@ fn ed25519_key_and_ed25519_detects_bad_signature() {
...
-        webpki::ECDSA_P384_SHA256,
-        webpki::ECDSA_P384_SHA384,
-        webpki::RSA_PKCS1_2048_8192_SHA256,
-        webpki::RSA_PKCS1_2048_8192_SHA384,
-        webpki::RSA_PKCS1_2048_8192_SHA512,
-        webpki::RSA_PKCS1_3072_8192_SHA384,
-        webpki::RSA_PSS_2048_8192_SHA256_LEGACY_KEY,
-        webpki::RSA_PSS_2048_8192_SHA384_LEGACY_KEY,
-        webpki::RSA_PSS_2048_8192_SHA512_LEGACY_KEY,
+        webpki::ring::ECDSA_P256_SHA256,
```
Comment (djc) [inline tests/signatures.rs]: Could/should these (and the other long lists in tests) use `ALL_`?

File: tests/signatures.rs [hunk 6]
```diff
@@ -35,7 +35,7 @@ fn ed25519_key_and_ed25519_good_signature() {
     let ee = include_bytes!("signatures/ed25519.ee.der");
     let message = include_bytes!("signatures/message.bin");
     let signature = include_bytes!("signatures/ed25519_key_and_ed25519_good_signature.sig.bin");
-    assert_eq!(check_sig(ee, webpki::ED25519, message, signature), Ok(()));
+    assert_eq!(check_sig(ee, webpki::ring::ED25519, message, signature), Ok(()));
```
Comment (djc) [inline tests/signatures.rs]: Maybe we should be importing these algorithms at the top of the module, so that the full paths aren't repeated in every test, but just once at the top? That could be a separate commit.

**Comment by cpu** [comment]
Could this be revived in a non-draft status now that the upstream fixed Windows support or are we waiting for a release tag there? :thinking: 

I think it's important to get this into the blocking list for https://github.com/rustls/webpki/issues/159

File: Cargo.toml [hunk 7]
```diff
@@ -66,10 +66,12 @@ name = "webpki"
 [features]
 default = ["std", "ring"]
 ring = ["dep:ring"]
+aws_lc_rs = ["dep:aws-lc-rs"]
```
[the remark this answers is not in the document]
Comment (ctz) [reply Cargo.toml:70]: have done that

File: tests/signatures.rs (hunk 5 above)
Comment (ctz) [reply tests/signatures.rs]: some of these lists are things like "all the algorithms except RSA_PSS_2048_8192_SHA256_LEGACY_KEY" which could be addressed by filtering the ALL_ list and removing the desired one? probably not going to do that in this PR, but will keep it in mind for a later refactor

**Comment by ctz** [comment]
nb. coverage job is broken due to nightly breaking `time`; see https://github.com/time-rs/time/issues/618

**Comment by ctz** [comment]
Apologies, CI is telling me this is not completely baked!

File: src/verify_cert.rs [hunk 8]
```diff
@@ -495,10 +495,9 @@ enum Role {
     EndEntity,
 }
 
-#[cfg(test)]
+#[cfg(all(test, feature = "alloc", feature = "ring"))]
```
Comment (cpu) [inline src/verify_cert.rs]: Thanks :+1: I also moved to a module-wide requirement for `alloc` in https://github.com/rustls/webpki/pull/179

File: src/test_utils.rs [hunk 9]
```diff
@@ -27,6 +27,7 @@ pub(crate) fn make_issuer(
     rcgen::Certificate::from_params(ca_params).unwrap()
 }
 
+#[cfg_attr(not(feature = "ring"), allow(dead_code))]
```
Comment (cpu) [inline src/test_utils.rs:30]: I think with the refactoring I did in https://github.com/rustls/webpki/pull/178 this will fall away because `end_entity.rs`'s `printable_string_common_name` test will use the `make_end_entity` helper from a test that doesn't require `ring` :crossed_fingers:

File: src/verify_cert.rs [hunk 10]
```diff
@@ -495,7 +495,7 @@ enum Role {
     EndEntity,
 }
 
-#[cfg(all(test, feature = "alloc", feature = "ring"))]
+#[cfg(all(test, feature = "alloc", any(feature = "ring", feature = "aws_lc_rs")))]
```
Comment (cpu) [inline src/verify_cert.rs:498]: nit: it would be nice to separate the `src/*.rs` & `tests/*.rs` updates from the `ci.yml` updates.

**Review by cpu — APPROVED** [review]

File: src/verify_cert.rs (hunk 10 above)
Comment (ctz) [reply src/verify_cert.rs:498]: have split this commit in two
